streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java

```diff
@@ -0,0 +1,230 @@
+            throw new IllegalArgumentException("storeName cannot be null");
+        }
+        final Set<String> sourceTopics = builder.stateStoreNameToSourceTopics().get(storeName);
+        if (sourceTopics == null) {
+            return Collections.emptyList();
+        }
+
+        final ArrayList<StreamsMetadata> results = new ArrayList<>();
+        for (StreamsMetadata instance : allMetadata) {
+            if (instance.stateStoreNames().contains(storeName)) {
+                results.add(instance);
+            }
+        }
+        return results;
+    }
+
+    /**
+     * Find the {@link StreamsMetadata}s for a given storeName and key.
+     * Note: the key may not exist in the {@link org.apache.kafka.streams.processor.StateStore},
+     * this method provides a way of finding which {@link StreamsMetadata} it would exist on.
+     *
+     * @param storeName     Name of the store
+     * @param key           Key to use to for partition
+     * @param keySerializer Serializer for the key
+     * @param <K>           key type
+     * @return The {@link StreamsMetadata} for the storeName and key
+     */
+    public synchronized <K> StreamsMetadata getMetadataWithKey(final String storeName,
+                                                               final K key,
+                                                               final Serializer<K> keySerializer) {
+        if (keySerializer == null) {
+            throw new IllegalArgumentException("keySerializer cannot be null");
+        }
+
+        // To find the correct instance for a given key we need to use the same partitioning
+        // strategy that was used when producing the messages. In this case we create a StreamPartitioner
+        // that works the same as org.apache.kafka.clients.producer.internals.DefaultPartitioner
+        return getMetadataWithKey(storeName, key, new StreamPartitioner<K, Object>() {
+            @Override
+            public Integer partition(final K key, final Object value, final int numPartitions) {
+                final String sourceTopic = builder.stateStoreNameToSourceTopics().get(storeName).iterator().next();
+                final byte[] bytes = keySerializer.serialize(sourceTopic, key);
+                return toPositive(Utils.murmur2(bytes)) % numPartitions;
+            }
+        });
+    }
+
+    /**
+     * Find the {@link StreamsMetadata}s for a given storeName and key.
+     * Note: the key may not exist in the {@link StateStore},
+     * this method provides a way of finding which {@link StreamsMetadata} it would exist on.
+     *
+     * @param storeName   Name of the store
+     * @param key         Key to use to for partition
+     * @param partitioner partitioner to use to find correct partition for key
+     * @param <K>         key type
+     * @return The {@link StreamsMetadata} for the storeName and key
+     */
+    public synchronized <K> StreamsMetadata getMetadataWithKey(final String storeName,
```

Comment:
One general observation: we can only control that `DefaultPartitioner` is used in the DSL layer and only for internal topics; for source topics of intermediate topics we cannot control which partitioner users will pick (note `though` etc can take `StreamPartitioner` as parameter), and also for the processor API, since repartitioning etc is also managed by users themselves the library does not control this as well.

So generally speaking it is not safe to query with a given key unless the user makes sure themselves that the `DefaultPartitioner` is used for writing to that topic which pipes to the store. For DSL we can mitigate this problem by remembering which partitioner class used except the source topics for determining the key; for processor API we cannot do much since library does not know the processor internals.